src/solc_0.8/defi/rules/RequirementsRules.sol

```diff
@@ -0,0 +1,323 @@
+//SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.2;
+
+import {Ownable} from "@openzeppelin/contracts-0.8/access/Ownable.sol";
+import {Address} from "@openzeppelin/contracts-0.8/utils/Address.sol";
+import {Math} from "@openzeppelin/contracts-0.8/utils/math/Math.sol";
+import {IERC721} from "@openzeppelin/contracts-0.8/token/ERC721/IERC721.sol";
+import {IERC1155} from "@openzeppelin/contracts-0.8/token/ERC1155/IERC1155.sol";
+
+contract RequirementsRules is Ownable {
+    using Address for address;
+
+    // maxStake amount allowed if user has no ERC721 or ERC1155
+    uint256 public maxStakeOverall;
+
+    struct ERC721RequirementRule {
+        uint256[] ids;
+        bool balanceOf;
+        uint256 minAmountBalanceOf;
+        uint256 maxAmountBalanceOf;
+        uint256 minAmountId;
+        uint256 maxAmountId;
+        uint256 index;
+    }
+
+    struct ERC1155RequirementRule {
+        uint256[] ids;
+        uint256 minAmountId;
+        uint256 maxAmountId;
+        uint256 index;
+    }
+
+    mapping(IERC721 => ERC721RequirementRule) internal _listERC721;
+    mapping(IERC1155 => ERC1155RequirementRule) internal _listERC1155;
+    IERC721[] internal _listERC721Index;
+    IERC1155[] internal _listERC1155Index;
+
+    event ERC11551ListRequirementSet(
+        address indexed contractERC1155,
+        uint256[] ids,
+        uint256 minAmountId,
+        uint256 maxAmountId
+    );
+    event ERC721ListRequirementSet(
+        address indexed contractERC721,
+        uint256[] ids,
+        bool balanceOf,
+        uint256 minAmountBalanceOf,
+        uint256 maxAmountBalanceOf,
+        uint256 minAmountId,
+        uint256 maxAmountId
+    );
+    event MaxStakeOverallSet(uint256 newMaxStake, uint256 oldMaxStake);
+    event ERC1155RequirementDeleted(address indexed contractERC1155);
+    event ERC721RequirementDeleted(address indexed contractERC721);
+
+    modifier isContract(address account) {
+        require(account.isContract(), "RequirementsRules: invalid address");
```

Comment:
invalid address => is not Contract ?

Reply:
done -> https://github.com/thesandboxgame/sandbox-smart-contracts/pull/605/commits/5c9ea8b3a12e5bf0488cae139893892c57bcb270